File: frontend/src/models/MarketStructureResult.ts
export type StructureTrend = "BULLISH" | "BEARISH" | "RANGE" | "UNKNOWN";

export type StructureBreakDirection = "BULLISH" | "BEARISH" | "NONE";

export type SwingImportance = "MAJOR" | "MINOR" | "INTERNAL";

export interface MarketStructureResult {
  valid: boolean;
  trend: StructureTrend;
  breakDirection: StructureBreakDirection;
  swingHigh: number | null;
  swingLow: number | null;
  bos: boolean;
  choch: boolean;
  mss: boolean;
  liquiditySweep: boolean;
  displacement: boolean;
  timeframeAligned: boolean;
  swingImportance: SwingImportance;
  passedRules: string[];
  failedRules: string[];
  reasons: string[];
  warnings: string[];
}

File: backend/app/engines/structure/MarketStructureResult.py
from __future__ import annotations

from typing import Literal
from typing import TypedDict


class MarketStructureResult(TypedDict):
    valid: bool
    trend: Literal["BULLISH", "BEARISH", "RANGE", "UNKNOWN"]
    breakDirection: Literal["BULLISH", "BEARISH", "NONE"]
    swingHigh: float | None
    swingLow: float | None
    bos: bool
    choch: bool
    mss: bool
    liquiditySweep: bool
    displacement: bool
    timeframeAligned: bool
    swingImportance: Literal["MAJOR", "MINOR", "INTERNAL"]
    passedRules: list[str]
    failedRules: list[str]
    reasons: list[str]
    warnings: list[str]
